File: leetcode_practice/leetcode_python3/28. 找出字符串中第一个匹配项的下标.py
#KMP算法
from typing import List
class Solution:
    def getNext(self, needle):
        next = [0] * len(needle)
        j = 0
        for i in range(1, len(needle)):
            while j > 0 and needle[i] != needle[j]:
                j = next[j - 1]
            if needle[i] == needle[j]:
                j += 1
            next[i] = j
        return next
    def strStr(self, haystack: str, needle: str) -> int:
        if not needle:
            return 0
        next = self.getNext(needle)
        j = 0
        for i in range(len(needle)):
            while j > 0 and haystack[i] != needle[j]:
                j = next[j - 1]
            if haystack[i] == needle[j]:
                j += 1
            if j == len(needle):
                return i - j + 1
        return -1
    
#（版本一）前缀表（减一）

class Solution:
    def getNext(self, next, s):
        j = -1
        next[0] = j
        for i in range(1, len(s)):
            while j >= 0 and s[i] != s[j+1]:
                j = next[j]
            if s[i] == s[j+1]:
                j += 1
            next[i] = j
    
    def strStr(self, haystack: str, needle: str) -> int:
        if not needle:
            return 0
        next = [0] * len(needle)
        self.getNext(next, needle)
        j = -1
        for i in range(len(haystack)):
            while j >= 0 and haystack[i] != needle[j+1]:
                j = next[j]
            if haystack[i] == needle[j+1]:
                j += 1
            if j == len(needle) - 1:
                return i - len(needle) + 1
        return -1
#（版本二）前缀表（不减一）

class Solution:
    def getNext(self, next: List[int], s: str) -> None:
        j = 0
        next[0] = 0
        for i in range(1, len(s)):
            while j > 0 and s[i] != s[j]:
                j = next[j - 1]
            if s[i] == s[j]:
                j += 1
            next[i] = j
    
    def strStr(self, haystack: str, needle: str) -> int:
        if len(needle) == 0:
            return 0
        next = [0] * len(needle)
        self.getNext(next, needle)
        j = 0
        for i in range(len(haystack)):
            while j > 0 and haystack[i] != needle[j]:
                j = next[j - 1]
            if haystack[i] == needle[j]:
                j += 1
            if j == len(needle):
                return i - len(needle) + 1
        return -1
#（版本三）暴力法

class Solution(object):
    def strStr(self, haystack, needle):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        m, n = len(haystack), len(needle)
        for i in range(m):
            if haystack[i:i+n] == needle:
                return i
        return -1    
#（版本四）使用 index

class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        try:
            return haystack.index(needle)
        except ValueError:
            return -1
#（版本五）使用 find

class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        return haystack.find(needle)
	

File: tests/c_programs/leetcode/28. 找出字符串中第一个匹配项的下标.c
#include <stdio.h>
#include <string.h>
void getNext(int* next, const char* s) {
    int j = 0;        // j 代表：前缀的末尾位置，也代表了当前最长相等前后缀的长度
    next[0] = 0;      // 只有一个字符时，没有前后缀，长度为 0

    // i 代表：后缀的末尾位置
    for (int i = 1; s[i] != '\0'; i++) {
        
        // 【情况 1】前后缀不匹配了
        // 这是一行最烧脑的代码：回溯。
        // 既然当前位置不匹配，我们就看前一个位置能跳到哪里，直到找着匹配的或者退回起点
        while (j > 0 && s[i] != s[j]) {
            j = next[j - 1]; 
        }

        // 【情况 2】前后缀匹配成功
        if (s[i] == s[j]) {
            j++; // 长度加 1
        }

        // 把计算出来的长度赋给 next 数组
        next[i] = j;
    }
}
int strStr(char* haystack, char* needle) {
    int h_len = strlen(haystack);
    int n_len = strlen(needle);
    if (n_len == 0){
        return 0;
    }
    int next[n_len];
    getNext(next, needle);
    int j = 0; // j 代表当前匹配到的 needle 的位置
    for (int i = 0; i < h_len; i++){
        while (j > 0 && haystack[i] != needle[j]){
            j = next[j - 1];
        }
        if (haystack[i] == needle[j]){
            j++;
        }
        if (j == n_len){
            return i-n_len + 1;
        }
    }
    return -1;
}